Fix unhandled exception in tiered compilation thread

janvorli opened the pull request:
The detection the unhandled exception is not working in the case of the tiered compilation thread, because there is no indication that the exception is going to be caught in the `CEEInfo::runWithErrorTrap`. So the EH assumes the exception is going to be unhandled and exits the process.
The unhandled exception message was not seen in the repro case of the issue likely because the tiered compilation thread console output didn't get flushed. When running under a debugger and stepping over the `InternalUnhandledExceptionFilter_Worker`, it was actually printed out.

The fix is to add `DebuggerU2MCatchHandlerFrame` to the `CEEInfo::runWithErrorTrap`. That allows the EH to see that the exception will be caught there and let it flow through the `CallDescrWorker` to reach that trap.

I have also taken this opportunity and renamed the confusing `invalidRevPInvoke` local variable.

Closes #116676

janvorli committed:
Fix unhandled exception in tiered compilation thread

The detection the unhandled exception is not working in the case of
the tiered compilation thread, because there is no indication that
the exception is going to be caught in the `CEEInfo::runWithErrorTrap`.
So the EH assumes the exception is going to be unhandled and
exits the process.
The unhandled exception message was not seen in the repro case
of the issue likely because the tiered compilation thread console
output didn't get flushed. When running under a debugger and stepping
over the `InternalUnhandledExceptionFilter_Worker`, it was actually
printed out.

The fix is to add `DebuggerU2MCatchHandlerFrame` to the
`CEEInfo::runWithErrorTrap`. That allows the EH to see that the
exception will be caught there and let it flow through the
CallDescrWorker to reach that trap.

Closes #116676

jkotas commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -10585,6 +10585,8 @@ bool CEEInfo::runWithErrorTrap(void (*function)(void*), void* param)
+    DebuggerU2MCatchHandlerFrame catchFrame(true /* catchesAllExceptions */);
+
 #if !defined(TARGET_UNIX)
 
```
Do we need to Pop the frame manually before the method exits?

AaronRobinsonMSFT replied:
Is there a reason the `DebuggerU2MCatchHandlerFrame`'s dtor doesn't do the `Pop()`?

janvorli replied:
In general, the explicit frames don't pop themselves in the destructor. I have actually added that for Unix long time ago as it was needed for some Unix specific cases of unwinding exception through native frames (https://github.com/dotnet/coreclr/pull/3250), but on Windows, we've stayed on the plan of explicit popping. And it is possible that with the new EH, it is not needed anymore.
Here are some reasons I was able to figure out:
* Some frames are created in functions that use SEH and that's incompatible with instances of objects with destructors in those functions. The `DebuggerU2MCatchHandlerFrame` is actually such a frame.
* Some frames are created by asm helpers, so there is no destructor being called for those. Having the pops explicit makes all frames to handle pops in a consistent manner (explicitly).

jkotas commented:
Should we add a regression test for this?

janvorli commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -10585,6 +10585,8 @@ bool CEEInfo::runWithErrorTrap(void (*function)(void*), void* param)
 
+    DebuggerU2MCatchHandlerFrame catchFrame(true /* catchesAllExceptions */);
+
 #if !defined(TARGET_UNIX)
```
It seems the Windows way works on Unix too, so I'll keep just that. This split is pre-.NET 1.0.

jkotas replied:
Can it be the other way around - keep the simpler Unix version? The fewer `PAL_TRY`s we have, the better.

AaronRobinsonMSFT commented on src/coreclr/vm/exceptionhandling.cpp:
```diff
@@ -3958,18 +3958,18 @@ extern "C" CLR_BOOL QCALLTYPE SfiNext(StackFrameIterator* pThis, uint* uExCollid
 
-        EH_LOG((LL_INFO100, "SfiNext: reached native frame at IP=%p, SP=%p, invalidRevPInvoke=%d\n",
-            GetIP(pThis->m_crawl.GetRegisterSet()->pCurrentContext), GetSP(pThis->m_crawl.GetRegisterSet()->pCurrentContext), invalidRevPInvoke));
+        EH_LOG((LL_INFO100, "SfiNext: reached native frame at IP=%p, SP=%p, isPropagatingToNativeCode=%d\n",
+            GetIP(pThis->m_crawl.GetRegisterSet()->pCurrentContext), GetSP(pThis->m_crawl.GetRegisterSet()->pCurrentContext), isPropagatingToNativeCode));
```
```suggestion
        EH_LOG((LL_INFO100, "SfiNext: reached native frame at IP=%p, SP=%p, isPropagatingToNativeCode=%s\n",
            GetIP(pThis->m_crawl.GetRegisterSet()->pCurrentContext), GetSP(pThis->m_crawl.GetRegisterSet()->pCurrentContext), (isPropagatingToNativeCode ? "true" : "false")));
```

janvorli committed:
Add regression test and fix GC mode

janvorli committed:
Make the regression test actually work

It needs to be compiled as optimized and the Console.WriteLine is needed to repro
the issue.

jkotas commented on src/tests/Regressions/coreclr/GitHub_116676/test116676.cs:
```diff
@@ -0,0 +1,29 @@
+    {
+        for (int i = 0; i < 1_000_000; i++)
+        {
+            Console.WriteLine(i);
```
This will produce megabytes of output that will get copied around, etc. Can we do something less visible, like `Thread.Sleep(10)` and change the for-loop to go till like 200? We should only need to hit the loop about `CallCountThreshold` times (30) to trigger the background tiered compilation.

janvorli replied:
I don't think it is needed, the test now repros the problem reliably (with the fix from this PR removed)

janvorli committed:
Modify the regression test to not to print to console

janvorli committed:
PR feedback

jkotas reviewed (approved):
Thanks

janvorli committed:
Update src/coreclr/vm/jitinterface.cpp

Fix contract

Co-authored-by: Jan Kotas <jkotas@microsoft.com>
